blog-posts/lambda-extensions/readme.md

```diff
@@ -0,0 +1,443 @@
+---
+published: false
+title: 'Power up your serverless application with AWS Lambda extensions'
+cover_image: 'https://raw.githubusercontent.com/CorentinDoue/articles/master/blog-posts/lambda-extensions/assets/cov.jpg'
+description: 'Implementation guide for AWS lambda extensions with NodeJs and Typescript'
+tags: aws, serverless, lambda, extensions
+series:
+canonical_url:
+---
+
+The AWS Lambda service has some killer features that you probably don't know about. One of them is the ability to run additional code alongside the handler code. This feature is called [extensions](https://docs.aws.amazon.com/lambda/latest/dg/lambda-extensions.html).
```

Comment:
AWS Lambda service comes packed with impressive features that might still be new to you. One such feature allows the running of supplementary code concurrently with the handler code, known as [extensions](https://docs.aws.amazon.com/lambda/latest/dg/lambda-extensions.html).

In this article, we'll delve into the process of using these extensions to create a personalized monitoring tool for lambdas without compromising their performance. The setup we'll use involves a NodeJs Lambda deployed with AWS CDK.

⚡ Delving into the Lambda Extension Feature
Let's begin with a brief overview of the Lambda extension. If you're already acquainted with this feature, feel free to skip ahead to the creation section.

🤔 What is a Lambda Extension?
A Lambda extension is a snippet of code executed in the same environment as the lambda handler. It can be:

Internal: This runs within the same process as the handler code.
External: This operates in a separate process.
lambda extensions schema

Regardless of the type, it shares the memory, timeout, and ephemeral storage of the execution environment with the handler.

The Lambda extension can interact with the Lambda service, getting notified about every event received by the handler and receiving telemetry events from the Lambda service.

An internal extension can directly enhance the handler code, given that it operates in the same process. For instance, an internal monitoring extension can supersede the HTTP library to intercept and log all HTTP calls.

Conversely, because an external extension operates in a separate process, it can enhance the handler without affecting its performance. For instance, an external monitoring extension can transmit metrics to an external monitoring service once the handler has responded, thereby avoiding any negative impact on API performance.

🧩 How to Implement a Lambda Extension
A Lambda extension is delivered in the form of a [Lambda layer](https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-concepts.html#gettingstarted-concepts-layer).

To integrate an external extension, you simply need to append the layer to your lambda. The extension process will then initiate automatically.

For an internal extension, while you still need to add the layer to your lambda, you'll also need to adjust the runtime to run the extension's entry point alongside the handler. This adjustment is accomplished via [runtime specific environment variables or wrapper scripts](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-modify.html).

For instance, in NodeJs, you can set the NODE_OPTIONS environment variable to --require /opt/my-extension.js or configure the AWS_LAMBDA_EXEC_WRAPPER environment variable to the path of a wrapper script.